### PR TITLE
api: GetObject caching implemented at XL layer.

### PR DESCRIPTION
**Pull request by harshavardhana**

### CODE REVIEW

File: pkg/cache/buffer.go [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package cache
```
Comment (abperiasamy) [inline pkg/cache/buffer.go]: name this package as ObjectCache

File: pkg/cache/buffer.go [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+	"bytes"
+	"errors"
+	"io"
+	"time"
+)
+
+// A Buffer is a variable-sized buffer of bytes with Read, Write and Seek methods.
+// The zero value for Buffer is an empty buffer ready to use.
+type Buffer struct {
+	buf        []byte    // contents are the bytes buf[off : len(buf)]
```
Comment (abperiasamy) [inline pkg/cache/buffer.go]: why not use Bytes.Buffer?

File: pkg/cache/cache.go [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+// NoExpiration represents caches to be permanent and can only be deleted.
+var NoExpiration = time.Duration(0)
+
+// Cache holds the required variables to compose an in memory cache system
+// which also provides expiring key mechanism and also maxSize.
+type Cache struct {
+	// Mutex is used for handling the concurrent
+	// read/write requests for cache
+	sync.Mutex
```
Comment (abperiasamy) [inline pkg/cache/cache.go]: this should be RWMutex.

**Comment by harshavardhana** [comment]
> why not use Bytes.Buffer?

bytes.Buffer doesn't provide Seeking.

File: pkg/objcache/objcache.go [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+		expiry:  expiry,
+	}
+}
+
+// ErrKeyNotFoundInCache - key not found in cache.
+var ErrKeyNotFoundInCache = errors.New("key not found in cache")
+
+// ErrCacheFull - cache is full.
+var ErrCacheFull = errors.New("cache is full no more entries allowed")
+
```
Comment (abperiasamy) [inline pkg/objcache/objcache.go:75]: "Not enough space"

File: xl-v1.go [hunk 5]
```diff
@@ -175,6 +162,8 @@ func newXLObjects(disks []string) (ObjectLayer, error) {
 		dataBlocks:    dataBlocks,
 		parityBlocks:  parityBlocks,
 		listPool:      newTreeWalkPool(globalLookupTimeout),
+		// Initialize object caching, FIXME: cache has no expiration for the time being.
+		objCache: objcache.New(maxCacheSize, objcache.NoExpiration),
 	}
```
Comment (abperiasamy) [inline xl-v1.go:172]: globalCacheSize should be initialized to maxCacheSize or MINIO_CACHE_SIZE environment variable.

File: xl-v1-errors.go [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package main
+
+import "errors"
+
+// errXLMaxDisks - returned for reached maximum of disks.
+var errXLMaxDisks = errors.New("Number of disks are higher than supported maximum count '16'")
```
Comment (abperiasamy) [inline xl-v1-errors.go:22]: Error message are inconsistently capitalized.

File: xl-v1-errors.go [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+// errXLMinDisks - returned for minimum number of disks.
+var errXLMinDisks = errors.New("Number of disks are smaller than supported minimum count '8'")
```
Comment (abperiasamy) [inline xl-v1-errors.go]: "Minimum '6' drives are required enable erasure code"

This 6 should be a global constant.

File: xl-v1-errors.go [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+// errXLNumDisks - returned for odd number of disks.
+var errXLNumDisks = errors.New("Number of disks should be multiples of '2'")
```
Comment (abperiasamy) [inline xl-v1-errors.go]: Total number of drives should be in multiples of '2'

File: xl-v1-errors.go [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+
+// errXLReadQuorum - did not meet read quorum.
+var errXLReadQuorum = errors.New("I/O error.  did not meet read quorum.")
```
Comment (abperiasamy) [inline xl-v1-errors.go]: Read failed. Insufficient number of drives online.

File: xl-v1-errors.go [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+
+// errXLWriteQuorum - did not meet write quorum.
+var errXLWriteQuorum = errors.New("I/O error.  did not meet write quorum.")
+
```
Comment (abperiasamy) [inline xl-v1-errors.go:35]: Write failed. Insufficient number of drives online.

File: xl-v1-errors.go [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+// errXLDataCorrupt - err data corrupt.
+var errXLDataCorrupt = errors.New("data likely corrupted, all blocks are zero in length")
```
Comment (abperiasamy) [inline xl-v1-errors.go]: We should revisit this case. This is unreliable error message.

**Pull request merged**